storage: don't clobber HardState on splits

tbg opened the pull request:
Since the move to proposer-evaluated KV, we were potentially clobbering the
HardState on splits as we accidentally moved HardState synthesis upstream
of Raft as well. This change moves it downstream again.

Though not strictly necessary, writing lastIndex was moved as well. This is
cosmetic, though it aids @irfansharif's PR #16809, which moves lastIndex to
the Raft engine. After this PR, neither HardState nor last index keys are
added to the WriteBatch, so that pre-#16993 `TruncateLog` is the only
remaining command that does so (and it, too, won't keep doing that for
long).

Note that there is no migration concern.

Fixes #16749.

tbg committed:
util: expose (and use) RaceEnabled boolean

This was previously in storage/engine, which is not a canonical place for it.

tbg committed:
storage: in race build, assert HardState integrity

Motivated by #16749. Added an assertion that catches HardState clobbering. Now

```
make stressrace PKG=./pkg/storage/ TESTS=TestStoreRangeSplitRaceUninitializedRHS
```

fails immediately with

```
clobbered hard state: [Term: 8 != 9 Commit: 10 != 0]

previously: raftpb.HardState{
    Term:             0x9,
    Vote:             0x2,
    Commit:           0x0,
    XXX_unrecognized: nil,
}
overwritten with: raftpb.HardState{
    Term:             0x8,
    Vote:             0x2,
    Commit:           0xa,
    XXX_unrecognized: nil,
}
```

which is fixed in the next commit in this PR.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17051)
<!-- Reviewable:end -->

irfansharif commented on pkg/storage/client_split_test.go:
```diff
@@ -2212,3 +2212,131 @@ func TestPushTxnQueueDependencyCycleWithRangeSplit(t *testing.T) {
 	}
 }
+
+func TestMinimal(t *testing.T) {
```
ha, `TestMinimal` was something I use for the quick hacky test. Can/Should be renamed (also breaks my hacky workflows).

tschottdorf replied:
Oops, didn't actually mean to commit. Removed.

tschottdorf commented:
Hmm, the assertion I put in is obviously garbage. Let me rework that.

tschottdorf commented:
Cleaned up, PTAL.

bdarnell commented:
Are there really no migration concerns? When this goes in, leaders will stop emitting HardStates in their write batches for splits, but followers may not yet know that they need to do it themselves. I don't think that will work because we won't fill in the missing HardState values at read time. 

---

Reviewed 5 of 14 files at r1, 1 of 2 files at r3, 8 of 8 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_state.go, line 489 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17051#-KpH5TzhtRx1aGN5f66f:-KpH5TzhtRx1aGN5f66g:bavxzsq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7db82c429f2fd21195ac03747bfb7d223e6a9ae0/pkg/storage/replica_state.go#L489)):*
> ```Go
> // synthesizeHardState synthesizes a HardState from the given ReplicaState and
> // any existing on-disk HardState in the context of a split, while verifying
> // that the application of the snapshot does not violate Raft invariants. It
> ```

The remaining mentions of `snapshot` in this comment look stale.

---

*[pkg/storage/replica_state.go, line 498 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17051#-KpH4U5Xg11MbtY2hVhH:-KpH4U5Xg11MbtY2hVhI:bay49kb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7db82c429f2fd21195ac03747bfb7d223e6a9ae0/pkg/storage/replica_state.go#L498)):*
> ```Go
> 	eng engine.ReadWriter,
> 	oldHS raftpb.HardState,
> 	truncState roachpb.RaftTruncatedState,
> ```

Instead of passing in truncState and raftAppliedIndex, maybe this should take `oldHS, newHS raftpb.HardState`. Or since this is only used for bootstrapping and splits, can we hard-code raftInitialLogTerm and raftInitialLogIndex? It doesn't look like we ever use this when the term and index might have other values. 

---

*[pkg/storage/client_split_test.go, line 2216 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17051#-KpGriRhtH6sAdHivJbk-r1-2216:-KpH16Km0FSImWG1r2fW:b-1idkuq) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfc176892bc2b756af4ffe377183d61a8a195ef1/pkg/storage/client_split_test.go#L2216)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Oops, didn't actually mean to commit. Removed.
</blockquote></details>

It's now added in one commit and removed in the next one. You can remove this file completely from both commits. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17051#-:-KpH5xrOEFxirMNHYYjl:b1v5c3q)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
You're right, my statement is wrong. We'll have to migrate as per the usual. I think I'll send a quick PR to get some stubs into place so that we can start landing changes before the migration RFC is implemented. Modified the commit message to point out the migration concern.

---

Review status: 3 of 11 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica_state.go, line 489 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17051#-KpH5TzhtRx1aGN5f66f:-KpHCyoDB9FwA5QppH9L:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7db82c429f2fd21195ac03747bfb7d223e6a9ae0/pkg/storage/replica_state.go#L489)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The remaining mentions of `snapshot` in this comment look stale.
</blockquote></details>

Done.

---

*[pkg/storage/replica_state.go, line 498 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17051#-KpH4U5Xg11MbtY2hVhH:-KpHDcEaPiad8C_oP1Yl:b5aggu8) ([raw file](https://github.com/cockroachdb/cockroach/blob/7db82c429f2fd21195ac03747bfb7d223e6a9ae0/pkg/storage/replica_state.go#L498)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Instead of passing in truncState and raftAppliedIndex, maybe this should take `oldHS, newHS raftpb.HardState`. Or since this is only used for bootstrapping and splits, can we hard-code raftInitialLogTerm and raftInitialLogIndex? It doesn't look like we ever use this when the term and index might have other values. 
</blockquote></details>

I like to avoid hard-coding `raftInitialLog{Term,Index}`. We don't use this code in much generality, but I prefer it as written, where you can (hope to) reason about its correctness without knowing the contexts in which it is called. I also don't mind that we pass a truncated state here as that is the source of the intial log term and index - it's a good reminder of how things tie together. In the absence of strong opinions, I'd just leave as is.

---

*[pkg/storage/client_split_test.go, line 2216 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17051#-KpGriRhtH6sAdHivJbk-r1-2216:-KpHCaKokrecrvKN3-UD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfc176892bc2b756af4ffe377183d61a8a195ef1/pkg/storage/client_split_test.go#L2216)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It's now added in one commit and removed in the next one. You can remove this file completely from both commits. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17051#-:-KpHEdyNpGZ7tUho1bDH:b-vin0j5)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Added the "migration" as well.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 2 of 12 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17051#-:-KpIQOYsEfsvpAaRAjMM:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 14 files at r1, 9 of 9 files at r5, 1 of 1 files at r6, 7 of 7 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/below_raft_protos_test.go, line 80 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/17051#-KpKIEcA5LZyVrn41G6M:-KpKIEcA5LZyVrn41G6N:bqjhqc0) ([raw file](https://github.com/cockroachdb/cockroach/blob/955042012b2cce6ca7d9d86f3c0e2d5d29a38b60/pkg/storage/below_raft_protos_test.go#L80)):*
> ```Go
> 		populatedConstructor: func(r *rand.Rand) proto.Message {
> 			n := r.Uint64()
> 			// NB: this would rot if HardState ever picked up more (relevant)
> ```

why not generate a new function like these other protos have?

---

*[pkg/storage/replica.go, line 4413 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/17051#-KpKHwQhOev-R_S4IJbc:-KpKHwQifoBeIU0I36Ty:b-lxayu4) ([raw file](https://github.com/cockroachdb/cockroach/blob/831d476ac6aa34af1c2694457e5774c9a6114f27/pkg/storage/replica.go#L4413)):*
> ```Go
> 		newHS, err := loadHardState(ctx, r.store.Engine(), rResult.Split.RightDesc.RangeID)
> 		if err != nil {
> 			panic(err)
> ```

is this intentionally mixing log.Fatal and panic?

---

*[pkg/storage/replica_command.go, line 3079 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/17051#-KpKItLkPNHRcFD5Wz1w:-KpKItLkPNHRcFD5Wz1x:bwct0p3) ([raw file](https://github.com/cockroachdb/cockroach/blob/955042012b2cce6ca7d9d86f3c0e2d5d29a38b60/pkg/storage/replica_command.go#L3079)):*
> ```Go
> 			// won't write a HardState at all and is guaranteed to crash.
> 			if err := makeReplicaStateLoader(split.RightDesc.RangeID).synthesizeRaftState(ctx, batch); err != nil {
> 				return enginepb.MVCCStats{}, EvalResult{}, errors.Wrap(err, "unable to write initial Raft state")
> ```

s/write initial/synthesize/?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17051)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: don't clobber HardState on splits

Since the move to proposer-evaluated KV, we were potentially clobbering the
HardState on splits as we accidentally moved HardState synthesis upstream of
Raft as well. This change moves it downstream again.

Though not strictly necessary, writing lastIndex was moved as well. This is
cosmetic, though it aids @irfansharif's PR #16809, which moves lastIndex to
the Raft engine. After this PR, neither HardState nor last index keys are
added to the WriteBatch, so that pre-#16993 `TruncateLog` is the only remaining
command that does so (and it, too, won't keep doing that for long).

Migration concerns: a lease holder running the new version will propose splits
that don't propose the HardState to Raft. A follower running the old version
will not write the HardState downstream of Raft. In combination, the HardState
would never get written, and would thus be incompatible with the
TruncatedState. Thus, while 1.0 might be around, we're still sending the
potentially dangerous HardState.

Fixes #16749.

bdarnell commented:
Review status: 3 of 12 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/below_raft_protos_test.go, line 80 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/17051#-KpKIEcA5LZyVrn41G6M:-KpL8X2Hd34yUHNJp1RU:b-9nphvk) ([raw file](https://github.com/cockroachdb/cockroach/blob/955042012b2cce6ca7d9d86f3c0e2d5d29a38b60/pkg/storage/below_raft_protos_test.go#L80)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Wouldn't that entail mucking with the vendored proto? I guess we could upstream the `gogoproto.populate` option, but ... ok, I'll try. Not holding up this PR though. Lmk if you were thinking something else.
</blockquote></details>

Instead of upstreaming a new gogoproto option, It would be simpler to use reflection on the HardState struct to make sure that it only has the expected fields.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17051)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 8 files at r4, 9 of 9 files at r11.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/below_raft_protos_test.go, line 80 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/17051#-KpKIEcA5LZyVrn41G6M:-KpNnZxdoVF5m4Z8eog5:b2yxz6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/955042012b2cce6ca7d9d86f3c0e2d5d29a38b60/pkg/storage/below_raft_protos_test.go#L80)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Instead of upstreaming a new gogoproto option, It would be simpler to use reflection on the HardState struct to make sure that it only has the expected fields.
</blockquote></details>

Ah, it's upstream. @bdarnell's suggestion SGTM.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17051)*
<!-- Sent from Reviewable.io -->
